### PR TITLE
[Feature] Add Certificate Authentication Support for WebCmdlets

### LINKED ISSUE #4544: Invoke-RestMethod and Invoke-WebRequest don't support client certificate authentication with certificate files
`Invoke-RestMethod` and `Invoke-WebRequest` requires the certificate specified by _-Certificate_ to be imported in the certificate store. Both methods do not allow using certificates from files. This breaks client certificate authentication on non-Windows platforms.

**Proposal**: Support the use of certificate files for the client authentication by adding a new parameter _-CertificateFile_ or by using the x509 object from `Get-PfxCertificate` directly.

This was tested on Ubuntu 16.04 with PowerShell 6.0.0-beta.5 and on Windows with PowerShell 5.0.

Ref: https://github.com/PowerShell/PowerShell/issues/473 and https://github.com/PowerShell/PowerShell/issues/1865

### PR DESCRIPTION
**Pull request by markekraus**
Fixes #4544

`HttpClientHandler` now supports client certificates. Enabling client certificates in Core for `Invoke-WebRequest` and `Invoke-RestMethod` makes it possible to do certificate authentication on Linux.

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: We already have a common module to add certificates here:
https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Security/certificateCommon.psm1

We shouldn't duplicate the code.

Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: @TravisEz13 I tested with that. But the cert in `New-GoodCertificate` does not contain the proper key usage to be used for SSL/TLS Auth. It is issued for only Document Encryption. I was afraid if I change the cert in `certificateCommon.psm1` I might be break tests that depend on that restriction. Thus, I added a new function in  `WebCmdlets.Tests.ps1` instead. What do you suggest?

Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Move the module to a common location and add your function to it, with descriptions of each cert (at least what you learned.)  At the minimum, file an issue that this work needs to be done to avoid future duplication of effort.  Even then, I would suggest adding the comments to each now.

**Comment by markekraus** [comment]
@TravisEz13 I think what may need to be done is issue a self-signed cert that has all the key usage and enhanced key usages required by all the tests, a kind of one-cert-to-rule-them-all and then also move the module to a common area etc. 

For now I have added comments.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (dantraMSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: I suggest you check to see if you can mock using WebRequest.RegisterPrefix instead of relying on an external site.  We find external sites make the tests unreliable.  See https://github.com/PowerShell/PowerShell/issues/2504

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (dantraMSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: Suggest explicitly setting ClientCertificateOptions to ClientCertificateOption.Manual.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Especially introducing a dependency on a new site.

Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: I agree. But I lack the skill to implement this. I would need guidance or assistance.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (markekraus) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: added.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: It seems @SteveL-MSFT have plans to add HTTPS support in our test `HttpListener.psm1`

Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: @iSazonov awesome. that would be better. It will probably also need  some way to ensure the client authentication cert is received by the listener too or to enforce certificate authentication on the listener so it can fail if one is not provided.

@dantraMSFT  I'm not entirely sure `WebRequest.RegisterPrefix` will work on Core. Unless I'm missing something, I don't think it would affect anything being done with the new classes under `System.Net.Http` in use in Core. I was able to [make a mockup](https://gist.github.com/markekraus/624c7a4969c013e66c357a2d8adeeafc) with it, but it only works on Windows PowerShell (tested on 5.1). Unless I'm going totally in the wrong direction?

Comment (dantraMSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: You're right, that alone won't solve it.  I also looked at doing a quick change to the the httplistener side but, as written, the client cert is not visible to the listener.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: @markekraus Can you address this feedback please?

Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: I did [here](https://github.com/PowerShell/PowerShell/pull/4546#issuecomment-321892193) and with [305c048](https://github.com/PowerShell/PowerShell/pull/4546/commits/305c048399d7a2119c63a717eb05bb312fe78a97). Was that insufficient?

Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Sorry missed that.

**Comment by adityapatwardhan** [comment]
@TravisEz13 @dantraMSFT Can you have another look please?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (dantraMSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: @markekraus - I don't have a quick solution to mocking the call; however, I still question the use of the link (https://prod.idrix.eu/secure/).  I found a reference to it on stackoverflow but I cannot find any information on the parent site indicating its for public use.  Have I missed something?

Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: I couldn't find anything written on the site. But, it appears that @idrassi is the owner of https://prod.idrix.eu/secure/ and was the same one who replied on that stackoverflow post with those links. I can try reaching out to them for written consent if needed.

Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Can we use local CI web server? https://www.appveyor.com/docs/deployment/agent/

Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: @iSazonov  I think you meant [This](https://www.appveyor.com/docs/services-databases/#internet-information-services). There would probably have to be a separate solution for AppVeyor and [Travis CI](https://docs.travis-ci.com/user/gui-and-headless-browsers/#Starting-a-Web-Server). It would probably be best if there was some kind of cross-platform solution that worked regardless of CI environment.

Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: @markekraus Have you looked to see if there is some Linux (so it would work on both platforms) docker container that would work?  Perhaps ASP.NET core in Linux should be familiar to most people using this platform.

Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: I think we need some explicit statement linked to the site that we can use the site in our tests.

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
We need to find some solution to deal with https://prod.idrix.eu/secure/

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: @TravisEz13 unfortunately, I'm a novice with Docker and have no experience with ASP.NET. I could look into it, but It will take a considerable amount of time to figure out what all needs to be done. I'm not opposed to looking into it, but, perhaps there is some way to engage someone else for help on this? I'm not really sure how to proceed.

Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Maybe open a tracking Issue and don't block the PR?

Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: I'm open to that, but I don't think we should be regularly 
 running the tests without doing one one the other either:

1. Get explicit permission to use the site for automated tested
1. find an alternative
1. mark the tests a pending until the issue is resolved.

Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: @TravisEz13 Ok. Let me look into docker/aspnetcore. If I can't make heads or tails of it, I will mark the test pending and create an issue. 

On the docker topic, would it be best to include a subfolder with a dockerfile and whatever assets are needed, have the test code run the docker build, run the container, do the relevant test, and tear down the container? Or or there some better way to do that? I imagine that could be a rather slow test.

Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: It would definitely make these slow tests and I think feature tests.  There are some tests that use docker here:  https://github.com/PowerShell/PowerShell/tree/master/test/docker/networktest   But those tests, are meant to be a different suite.

I think @iSazonov  might be right.  I think option 3 might be the best option.  Mark the tests as pending, and file an issue.  If you are still willing to work on trying to fix the tests, I'd love to work with you.  Assuming @adityapatwardhan (the maintainer for the PR) agrees.

Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Oh, and add a comment with the issue number saying why the tests are marked pending.

Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: @TravisEz13 Issue #4609 created, tests set to pending, and comments added. I'm still interested in figuring this out, I'm just unsure of the time it will take me to get up to speed on aspnetcore enough to do this.

**Review by TravisEz13 — APPROVED** [review]
I'm ok with this assuming Aditya agrees

**Review by adityapatwardhan — APPROVED** [review]
LGTM

**Comment by adityapatwardhan** [comment]
Had a chat with @TravisEz13 about the testing challenges. Now that #4609 is opened, I can merge this.

**Pull request merged**
